spec/models/principal_spec.rb

```diff
@@ -202,4 +202,325 @@
       expect(Firm.where(id: firm.id)).to be_empty
     end
   end
+
+  describe '#next_onboarding_action' do
+    context 'when principal has no parent firm' do
+      before :each do
+        principal.firm.destroy
```

Comment:
A principal with a nil/destroyed parent firm is not a real state in the system at the moment. The `after_create  :associate_firm` handler in principal.rb guarantees there is a skeleton firm from the moment it's created.

So I'm not sure if we should be handling this scenario like it's a valid state. If this were to happen it means something wrong has happened or the system has stopped behaving consistently.